proposed/2021/NuGetConsumePDBInPackageReference.md

```diff
@@ -0,0 +1,212 @@
+
+```json
+{
+  "version": 3,
+  "targets": {
+    ".NETCoreApp,Version=v5.0": {
+      "PackageA/1.0.0": {
+        "type": "package",
+        "compile": {
+           "lib/netstandard2.0/PackageA.dll": {
+              "related": "pdb,xml"
+           }
+        },
+        "runtime": {
+           "lib/netstandard2.0/PackageA.dll": {
+              "related": "pdb,xml"
+           }
+      }
+    }
+  },
+```
+#### Project reference:
+For example, one project has a project reference of ProjectA. Supppose in folder ./ProjectA/lib/net5.0, there are PackageA.pdb, PackageA.xml file besides ProjectA.dll.
+There will be no change in targets section in assets file for project reference:
+
+```json
+{
+  "version": 3,
+  "targets": {
+    ".NETCoreApp,Version=v5.0": {
+        "ClassLibrary1/1.0.0": {
+        "type": "project",
+        "framework": ".NETCoreApp,Version=v5.0",
+        "compile": {
+          "bin/placeholder/ClassLibrary1.dll": {}
+        },
+        "runtime": {
+          "bin/placeholder/ClassLibrary1.dll": {}
+        }
+      }
+    },
+```
+
+
+
+On the library side there are no changes. 
+
+This solution is adding a `related` property in [LockFileItem](https://github.com/NuGet/NuGet.Client/blob/4fef99532f4022504feec5f68c8501cbeadd3aed/src/NuGet.Core/NuGet.ProjectModel/LockFile/LockFileItem.cs). The [LockFileItem](https://github.com/NuGet/NuGet.Client/blob/4fef99532f4022504feec5f68c8501cbeadd3aed/src/NuGet.Core/NuGet.ProjectModel/LockFile/LockFileItem.cs) type which represents an element in the compile list already has a collection of properties. 
+
+Specifically:
+
+```cs
+  public string Path { get; }
+
+  public IDictionary<string, string> Properties { get; } = new Dictionary<string, string>();
+```
+
+So there is no need to change any code references LockFileItem. 
+
+When generating the targets section of assets file, NuGet needs to go through the runtime and compile time folder for the files with specific extensions and add those as the `related` property.  
+
+After it's generated, NuGet will not use the `related` property. It will be consumed at build time by .NET SDK. 
+
+In order to consume the `related` property, the .NET SDK will need to make change in the following part: 
+
+[dotnet/sdk/10947](https://github.com/dotnet/sdk/issues/10947) The build tasks for .NET Core SDK  [code](https://github.com/dotnet/sdk/blob/master/src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs)
+
+
+## Open Questions
+
+* In which version will this functionality be ready?
+
+* When there is no other extensions, shall NuGet provide empty string `"related": ""`, or just skip adding `"related": ""` in assets file?
+
+* Shall NuGet list not only .pdb and .xml, but also all the extensions in [AllowedReferenceRelatedFileExtensions](https://github.com/dotnet/msbuild/blame/main/src/Tasks/Microsoft.Common.CurrentVersion.targets#L621-L627) if there is any? (.pdb, .xml, .pri, .dll.config, .exe.config)
```

Comment:
I think `AllowedReferenceRelatedFileExtensions` should not be something NuGet reads, as it's SDK/project specific and it would make it more challenging for NuGet to cache assets in a performant way. 

As far as, which extensions NuGet checks for, I think the answer could be any and all. 

So if say there's a `a.randomext` next to `a.dll`, NUget would specify `randomext`.

Reply:
Thanks for the great suggestion! Applied.